### ISSUE TITLE
I can't reproduce results

### ISSUE THREAD

**Issue by wesleylp** [facebookresearch/maskrcnn-benchmark#376]
## ❓ Questions and Help
I'm trying to reproduce results for the same experiment.

I've tried everything discussed [here](https://github.com/pytorch/pytorch/issues/7068) but nothing worked.

I added the following lines of code in `tools/train_net.py`:
```python
+torch.backends.cudnn.benchmark = False
+torch.backends.cudnn.deterministic = True
+
+torch.manual_seed(42)
+torch.cuda.manual_seed(42)
+torch.cuda.manual_seed_all(42)
+random.seed(42)
+np.random.seed(42)
```
Also added the following in `maskrcnn_benchmark/data/build.py`
```python
data_loader = torch.utils.data.DataLoader(
            dataset,
            num_workers=num_workers,
            batch_sampler=batch_sampler,
            collate_fn=collator,
+           worker_init_fn=np.random.seed(42),
        )
```
To see if it works, I train for only180 iterations.

any clue?

**Comment by fmassa** [comment]
Ok, I re-read your question again and I think I understand what you meant (desculpa!).

one thing to modify the the `worker_init_fn`, which is not really performng what you want, check https://github.com/pytorch/pytorch/issues/7068#issuecomment-391389505

Apart from that, there are a few functions (including ROIAlign) which use atomic adds, and thus they are not guaranteed to give the exact same results over different runs.

**Comment by wesleylp** [comment]
Thank you for your answer! 

You don't have to apologize, I may have caused some confusion!
With 'reproduce results' I mean get the same (or quiet near) results if ran the same experiment twice.


I'm not sure if I understand [pytorch/pytorch#7068 (comment)](https://github.com/pytorch/pytorch/issues/7068#issuecomment-391389505) pretty well.
I tried the following in `maskrcnn_benchmark/data/build.py` (ignoring `worker_id`):
```python
data_loader = torch.utils.data.DataLoader(
            dataset,
            num_workers=num_workers,
            batch_sampler=batch_sampler,
            collate_fn=collator,
+           worker_init_fn=_init_fn(),
        )
. . .

+ def _init_fn():
+     np.random.seed(42)
```


Should I still be able to get similar results, even with the atomic operations?
They look quiet different, at my point of view:

- 1st round:
 ```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.344
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.634
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.306
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.262
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.476
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.111
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.382
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.382
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.295
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.524
 ```

- 2nd round:
 ``` 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.334
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.635
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.278
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.252
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.472
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.106
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.378
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.378
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.293
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.516
 ```

**Comment by fmassa** [comment]
You should pass `_init_fn` instead of `_init_fn()` to `worker_init_fn`, that's already a problem, because you are currently passing `None` to it (the return value of your function is None).

Because of the atomic operations, the results will not match exactly, and might diverge over time.

Here is what I propose: run one forward/backward iteration on a single GPU, and check that the outputs/gradients/weights are the same.
Also check that the examples that your batch is being fed to are the same.

This should indicate if the culprit is the atomic operations or not

**Comment by wesleylp** [comment]
Now, I've tried:

```python
data_loader = torch.utils.data.DataLoader(
            dataset,
            num_workers=num_workers,
            batch_sampler=batch_sampler,
            collate_fn=collator,
+           worker_init_fn=_init_fn,
        )
. . .

+ def _init_fn(worker_id):
+     np.random.seed(42 + worker_id)
```


But still get very different results:
- 1st round
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.339
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.635
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.319
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.257
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.105
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.376
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.376
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.291
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.515


```

- 2nd round:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.324
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.634
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.262
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.241
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.468
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.103
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.373
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.373
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.283
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.519

```

**Comment by wesleylp** [comment]
I'll try that and let you know!

**Comment by wesleylp** [comment]
After 1 iteration on a single GPU, I've got:

- 1st round
```
img_idx 157 80 3 141
2019-01-23 14:05:20,966 maskrcnn_benchmark.trainer INFO: eta: 0:00:00  iter: 1  loss: 1.3640 (1.3640)  loss_classifier: 0.6450 (0.6450)  loss_box_reg: 0.0002 (0.0002)  loss_objectness: 0.7127 (0.7127)  loss_rpn_box_reg: 0.0062 (0.0062)  time: 1.4850 (1.4850)  data: 0.9323 (0.9323)  lr: 0.001667  max mem: 4420

TRAIN RESULTS:
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.010
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.007
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.011
OrderedDict([('bbox', OrderedDict([('AP', 5.720460257359246e-05), ('AP50', 0.0003551128270798423),('AP75', 2.3630047968997378e-05),  ('APs', -1.0), ('APm', 6.805516953882868e-06), ('APl', 0.00014271250708355944)]))])

TEST RESULTS:
DONE (t=0.04s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.003
OrderedDict([('bbox', OrderedDict([('AP', 6.244442743448054e-06), ('AP50', 5.28911756986547e-05), ('AP75', 0.0), ('APs', -1.0), ('APm', 5.215468285094777e-06), ('APl', 9.385520529081461e-06)]))])
```

- 2nd round:
```
img_idx 157 80 3 141
2019-01-23 14:08:31,552 maskrcnn_benchmark.trainer INFO: eta: 0:00:00  iter: 1  loss: 1.3640 (1.3640)  loss_classifier: 0.6450 (0.6450)  loss_box_reg: 0.0002 (0.0002)  loss_objectness: 0.7127 (0.7127)  loss_rpn_box_reg: 0.0062 (0.0062)  time: 1.3627 (1.3627)  data: 0.8487 (0.8487)  lr: 0.001667  max mem: 4420

TRAIN RESULTS:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.010
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.007
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.011
OrderedDict([('bbox', OrderedDict([('AP', 5.720460257359246e-05), ('AP50', 0.0003551128270798423), ('AP75', 2.3630047968997378e-05), ('APs', -1.0), ('APm', 6.805516953882868e-06), ('APl', 0.00014271250708355944)]))])


TEST RESULTS:
DONE (t=0.04s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.003
OrderedDict([('bbox', OrderedDict([('AP', 6.244442743448054e-06), ('AP50', 5.28911756986547e-05), ('AP75', 0.0), ('APs', -1.0), ('APm', 5.215468285094777e-06), ('APl', 9.385520529081461e-06)]))])
```

The images in my batch are the same.
The losses seems the same .
Also, the results looks the same (I'm not sure I can say that as they are so close to 0).

**Comment by fmassa** [comment]
I think instead of observing the final results (which are averaged over a lot of images) you should be a bit more intrusive on the codebase and start an `IPython` interpreter (via `from IPython import embed; embed()`) and compare the outputs after a single iteration. It will make things easier to understand.

**Comment by wesleylp** [comment]
Sure, I will try that!
